### DISCUSSION TITLE
Room Blueprints for the Map Editor

### DISCUSSION THREAD

**Discussion by lloyd-lejack** [CorsixTH/CorsixTH#1967] [Research Room]
Hi, Lloyd here. I wanted to share something that came to my mind when i was designing the plot layout for a new level. 
I was thinking that it would be nice to have a blueprint feature where you can see the minimum size of a specific room such as the GP or the OP Theatre. This would help to check if your plot design has an optimal size that leaves space for corridors, or in case you want to fit a specific room in it without having to memorize the room sizes (both minimum and optimal) or going in-game just to check if your plot has the right size.
As i told Lewri, it's not really necessary to have prefab room, just the blueprint, unless you really wanted to ensure the furniture is distributed in the best way possible, which is too specific for a feature like this.

**Comment by TheCycoONE** [comment]
Would you be able to mock what it should look like? I think it would help a lot with having everyone on the same page.

**Comment by lloyd-lejack** [comment]
https://prnt.sc/177z092
The idea is to display the mininum size required and a suggested "optimal" size to build a room with by adding a blue square (that can only be placed in hospital tiles). 
This would avoid having to quit the editor to check the sizes in-game, since i'm sure not everyone is willing to memorize or write down each one in a notepad and sometimes you may want to double check if there's enough space in the corridors or if your map is able to fit all the rooms you have available.
It might not sound that useful with smaller rooms, but it's sometimes hard to design a plot to specifically fit the X-Ray, Ward and Operating Theater without making the plot too big or too small.
The HUD shouldn't be that complex, it would be split into the same categories as the build room feature. Buttons could use the same style as the tile HUD. If this feature is added, i think it would be nice to have a button to enable/disable each window so they doesn't occupy the entire screen (which would be a concern in laptops or PC with small monitors), this option would go on the upper menu, between the file and player chart menus (Alternatively this new menu could merge with the player chart as a "Map options" menu or something)
Sorry if there's any typo, it's 1 am here and i'm not double checking my spelling.

**Comment by Alberth289346** [comment]
Copied the image from the external provider, and converted to 8bpp,  so it stays available.

![roomsize2](https://user-images.githubusercontent.com/3254232/123654780-b31c3e80-d82e-11eb-8905-853f7a12726d.png)

**Comment by lloyd-lejack** [comment]
Thank you

**Comment by TheCycoONE** [comment]
hmm... I'm not sure we have a concept of an 'optimal size' today,  but if people agree on standard (square) sizes it could be added.

**Comment by mugmuggy** [comment]
On the Playstation version, when you go to build a room in game, it instantiates a blue print (the build walls phase) to the minimum size that you drag around, its fairly handy on that platform and I've thought about something like that would be handy.

As for what could constitute optimal size, standard square size, that allows multiple configurations in a room, which most of the time is the same as the minimum size and normally would be one extra tile for most other rooms. There are some rooms that once you place a door the ability to rearrange the room requires a rebuild, especially once you place a plant or extinguisher.

If then its minimum_size + 1, then we have a standard set of foot prints for all rooms as 4x4, through to 7x7. Ward, training and staff, toilets and research are the ones that have no real optimal size in that sense as they can be expanded to different sizes depending on the user and the content they wish to put in the room and the number of staff/patients they intend to use that room.

So could expand the optimal to include an 'all' size for a room size that allows multiple layouts with one of all possible room content added. But I suspect the majority would fit in the minimum_size + 1 category, but you might have to go + 2 to allow multiple door positions depending on orientations.

Someone else playing the game might have a completely different idea of what they want than what you think they might use that space for.

**Reply by TheCycoONE** [reply]
we are not going to compute the sizes, someone can go to each room and enter it as a property under `room.minimum_size` using whatever criteria you want - or entirely arbitrary; or skip it and just blueprint the minimum.

**Reply by Alberth289346** [reply]
We could copy minimum size to the configuration file as optimum size, then a user can edit that file.

I do think its value is rather limited, as different users have different ideas about optimum values.

**Reply by RetroGamerFan** [reply]
Optimal layout also needs to consider corridor space and how patients move from one room to the next and the placement of objects etc..

Here's a map I made trying to make everything optimal...

![image](https://user-images.githubusercontent.com/95074358/143554486-682f3214-6448-40f5-95c2-c284d0e2d9bf.png)

**Reply by Alberth289346** [reply]
You optimized for easy layout. That works well if you have plenty of room. However if you have a different level that has very little space for building, you cannot afford to waste any tile so this layout is not optimal in that level, you want as small rooms as possible or some rooms cannot be placed.

This is the whole problem with "optimal". There is no single roomsize that is optimal for all possible levels.

**Reply by RetroGamerFan** [reply]
The Map designer has more than enough room for creating large hospitals with an easy layout.

Figuring out the best layout for rooms for each level is a large part of the game. It's also a part of this remake that differs a lot from the original game too.

I would have to play the game again from the start if you want screenshots of what I consider to be the best layout for each level, but I'll do it if you insist?

**Reply by lewri** [reply]
I think a point missed here is not every designer wants a large hospital. Some like giving people the challenge of trying to maximize the most of your space i.e. small hospitals, windy corridors etc.

Theme hospitals original campaign was overly generous on space, too much so perhaps. Probably because they expected you to have bad maintenance during earthquakes and wanted the levels to remain passable even in disarray.

In CorsixTH you can now removed destroyed rooms if you enable the setting, meaning you don't need do much generosity in space anymore.

**Comment by RetroGamerFan** [comment]
Main Campaign - Best Layout - Level 01  - Toxicity

![image](https://user-images.githubusercontent.com/95074358/143867096-efee5f1b-d08b-49ef-a4db-7498a9fa6c45.png)

1. The GP's office closest to the entrance queue size is set to 2. Same for the Psychiatrist room closest to the GP offices.

2. The Ward and Pharmacy are placed close together since both are used by nurses.

3. The General Diagnosis is typically the next diagnosis room used after the GP's office.

4. The staff room is "centralised".

5. Rooms "fit" against the hospital wall layout for simplicity.

6. Radiators are centralised in each room.

7. Radiators outside of rooms are 6 or 7 tiles apart.

8. Drink machines are spaced every 3 or 4 tiles apart.

9. Benches are placed outside rooms against the walls so patients and staff can easily walk between rooms.

----------------------------------------------------------------------------------------------------

1. Should there be another staff room closer to the ward?

2. Should the toilets be placed by the entrance at the lower corner so patients can walk between rooms easier?

3. Should the staff room be placed in the middle of the main area and the 2 psychiatrist office made larger? (Keep in mind on level one there are not enough patient arrivals to warrant more than 2 GP offices with the hospital policy for tiredness set to 40%.

4. Should the door to the toilets be placed on the opposite side?

----------------------------------------------------------------------------------------------------

1. Mandela tells me benches next to toilets increase the chance of vomiting for any patients using them.

2. I am assuming book shelves and skeletons make the Physiatrist rooms more efficient like they do for training rooms.

**Comment by RetroGamerFan** [comment]
A revised layout which would make more sense if the first level had more patients on the hardest difficulty setting.

![image](https://user-images.githubusercontent.com/95074358/143872111-ef51eb14-8a99-47e3-8f54-6ac6bc260a4a.png)

Keep in mind the Ward layout makes no sense at all since no more than 2 beds are necessary for this level. It's all about reducing wasted space for the likes of janitors to occupy.

1. Additional GP's office for patients to visit after the Ward.

2. Additional staff room closer for the nurse in the ward to use.

3. Perhaps the GP's office and staff room should switch places?

4. Toilet placement is probably bad here really, since the patient has a longer distance to walk to and from any room they are visting.

Funny, the first level has a helipad, yet there are no emergencies. The second level has no helipad.

**Comment by RetroGamerFan** [comment]
Level 02 - Sleepy Hollow.

![image](https://user-images.githubusercontent.com/95074358/144225870-dfda99ff-8a49-4f15-891a-ee1e440639d3.png)

1. Due to money constraints the hospital layout went through several iterations.

2. The research room was removed (and accompanying staff fired), once all research had been completed.

3. Not happy with 3 tile width corridors, it gets too cramped, but since doctors using a 4x4 slack tongue clinic would gradually get stressed I made the room 4 x5 instead.

4. I was tempted to buy the larger plot instead of the smaller one due to the large amount of additional space and use that for the Ward, Pharmacy and the 2 Clinics. The main plot would then also have enough room to keep corridors 4 tiles apart, however, doctors would sometimes need to travel between plots. It's always a trade off with these kinds of maps.

**Comment by lewri** [comment]
So, in a game that I used to play a lot (Prison Architect) it had the option of being able to lay down prefabs of rooms. The prefabs generally allowed for slightly more space than necessary to improve happiness.

Perhaps we shouldn't be focussing on optimal size, but instead a prefab of a room layout where everything would comfortably fit with a min 1 tile gap between objects,

**Comment by Alberth289346** [comment]
Maybe just add a way to drag an area of the size you want, as long as it's recognizable from empty space and other such areas, it should be fine for rough measurements. Displaying size and name of a room that's supposed to go there would be helpful.

Can such information be stored in the current map file format? That would be useful.

It feels like a step towards the ability for making test-scenes more easily to me. That however would probably eventually mean to consider the map editor as a stripped version of the test-scene editor.

**Reply by RetroGamerFan** [reply]
Two Point Hospital allows you to save and load "Room Templates".
